pkg/tinyvcdiff/contrib/tinyvcdiff_vfs/vfs.c

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (C) 2022 Juergen Fitschen
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+#include "vcdiff_vfs.h"
+#include <unistd.h>
+#include "errno.h"
+
+static int _read (void *dev, uint8_t *dest, size_t offset, size_t len)
+{
+    int rc;
+    vcdiff_vfs_t *vfs = dev;
+
+    /* seek to the given offset */
+    rc = vfs_lseek(vfs->fd, offset, SEEK_SET);
+    if (rc < 0) {
+        return rc;
+    } else if ((size_t) rc != offset) {
+        return -EFAULT;
+    }
+
+    /* write until all bytes are stored in the buffer */
+    while (len) {
+        rc = vfs_read(vfs->fd, dest, len);
+        if (rc < 0) {
+            return rc;
+        }
+        len -= rc;
+        dest += rc;
+    }
```

Comment:
tbh I'm wondering why this is needed in the first place.
If the underlying filesystem driver would return less bytes than requested with the expectation that you call read again, that should be handled inside `vfs_read()`.

But normally if you read less bytes than you requested, that means the file just ended.

Reply:
Good point. At least I should break if `rc == 0` and return an error.

Aside from that, I'm not sure if `vfs_read()` is guaranteed to read all available data in one go. That's the reason for the loop. Do you have a hint if this is too careful?

Reply:
> I'm not sure if `vfs_read()` is guaranteed to read all available data in one go.

It should because it can grantee that.